bolt/lib/Rewrite/RewriteInstance.cpp

```diff
@@ -1276,6 +1280,57 @@ void RewriteInstance::discoverFileObjects() {
   registerFragments();
 }
 
+Error RewriteInstance::discoverFiniAddress() {
```

Comment:
Probably we should move this code and updateFiniReloc to InstrumentationRuntimeLibrary or even better to runtimeLibrary, since any of them might using this hack and it would be great to make it more universal, what do you think?

Reply:
afaict, the code is already universal as it only uses the `RuntimeLibrary` interface so it should work for any runtime library implementation. Or am I missing something?

Reply:
I mean that this code is runtimelibrary-specific, so probably it is better to move to runtimeLibrary

Reply:
I'd say it's more ELF-specific so `RewriteInstance` would be a better place for it. The logic only deals with somehow overwriting the fini address and, in principle, it doesn't really care about the runtime library. Currently, only the runtime library uses this feature but maybe it could be used by something else as well in the future?

Reply:
But this one doesn't overwrite it, it's only returns if the binary has DT_FINI or DT_FINI_ARRAY. Theoretically, of course, it might be used by something else, although I doubt to be honest. Also the mirror rewriting function "updateFiniReloc" is clearly runtime-specific due to getRuntimeLibrary call. Since these are 2 in pair for single purpose my humble opinion is that their place is in runtime-specific code..

Reply:
I feel like the one call to `getRuntimeLibrary` does not warrant moving those functions to the runtime library. But maybe we should abstract-away that call into something like `RewriteInstance::getFiniAddress` so that the logic here doesn't need to know about runtime libraries at all.

But the main reason for me is still that all this logic is very ELF-specific and would be completely different on Mach-O, for example. So implementing this in `RewriteInstance` and `MachORewriteInstance` seems to make more sense to me. Moreover, the logic is not only contained in `discoverFiniAddress` and `updateFiniReloc`; part of it is also in `readELFDynamic` and, if you consider the old logic of just updating `DT_FINI` as part of the same feature (which I do), in `patchELFDynamic`. The handling of `DT_FINI` in the latter two functions could also be seen as runtime library specific but moving them to the runtime library does not seem logical or even possible.

The way I see it, the runtime library simply wants to say "make sure this address is called after main"; it's up to the implementation of a particular object format (i.e., `RewriteInstance` or `MachORewriteInstance`) to make sure that actually happens.

Reply:
Agreed, done!